sdk/cpp/agent.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "absl/status/status.h"
+#include "absl/time/time.h"
+
+namespace ubbagent {
+
+struct AgentStatus {
+    absl::Time last_report_success;
+    int current_failure_count;
+    int total_failure_count;
+    absl::Status status;
+};
+
+
+class Agent {
+  public:
+    // Factory method to create Agent. Result of the operation will be written to out_status.
+    static std::unique_ptr<Agent> Create(const std::string& config, const std::string& state_dir, absl::Status* out_status);
+    
+    // Desctuctor shuts down the agent.
+    ~Agent();
```

Comment:
Shouldn't we have a separate `Shutdown` method to avoid doing this much work in the destructor?

Reply:
Switched from a destructor to a separate Shutdown method.